src/pushed_authorization.rs

```diff
@@ -0,0 +1,297 @@
+use std::collections::HashMap;
+
+use crate::{
+    authorization::AuthorizationDetail,
+    core::profiles::CoreProfilesAuthorizationDetails,
+    profiles::AuthorizationDetaislProfile,
+    proof_of_possession::{
+        ConversionError, ParsingError, ProofOfPossessionBody, VerificationError,
+    },
+};
+use oauth2::{AuthUrl, CsrfToken, PkceCodeChallenge};
+use openidconnect::{
+    core::CoreErrorResponseType, ClientId, IssuerUrl, RedirectUrl, StandardErrorResponse,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use serde_with::{json::JsonString, serde_as};
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ParRequestUri(pub String);
+
+impl ParRequestUri {
+    pub fn new_random(num_bytes: u32) -> Self {
+        use base64::{
+            alphabet,
+            engine::{self, general_purpose},
+            Engine as _,
+        };
+        use rand::{thread_rng, Rng};
+
+        const CUSTOM_ENGINE: engine::GeneralPurpose =
+            engine::GeneralPurpose::new(&alphabet::URL_SAFE, general_purpose::NO_PAD);
+        let random_bytes: Vec<u8> = (0..num_bytes).map(|_| thread_rng().gen::<u8>()).collect();
+        Self(format!(
+            "urn:ietf:params:oauth:request_uri:{:}",
+            CUSTOM_ENGINE.encode(random_bytes)
+        ))
+    }
+
+    pub fn get(&self) -> &String {
+        &self.0
+    }
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub struct Request {
+    redirect_uri: RedirectUrl,
+    client_id: ClientId,
+    #[serde_as(as = "JsonString")]
+    #[serde(rename = "authorization_details")] // TODO unsure what to do with it
+    authorization_details: Vec<AuthorizationDetail<CoreProfilesAuthorizationDetails>>,
+    state: CsrfToken,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    request_uri: Option<String>,
+    client_assertion: String,
+    client_assertion_type: String,
```

Comment:
AFAIK `client_assertion` isn't required for PAR.

Reply:
Yes, you are right. They say 

> The request also includes, as appropriate for the given client, any additional parameters necessary for client authentication (e.g., client_secret or client_assertion and client_assertion_type)

I'll make it optional.